### PR TITLE
Clean out .kibana_* indexes before restarting kibana

### PR DESCRIPTION
**Pull request by cachedout**
This resolves [the following error](https://internal-ci.elastic.co/job/elastic+estf-monitoring-snapshots+7.4+multijob-kibana/78/consoleFull):

`ERROR: Metricbeat-indexed doc for type='kibana_stats' has unexpected deletion: kibana_stats.usage.visualization_types`

The above error also [appeared recently here](https://internal-ci.elastic.co/job/elastic+estf-monitoring-snapshots+7.4+multijob-kibana/68/console).

I was receiving this error consistently in my testing on the 7.4 branch but after applying this fix, I can no longer produce the error after multiple tests runs.

I have a hunch (but cannot definitively prove) that this may resolve [other flaky tests](https://internal-ci.elastic.co/job/elastic+estf-monitoring-snapshots+7.4+multijob-kibana/74/console) wherein `kibana_stats` documents do not contain certain `usage` fields.

Related to the reason that this seems to happen is that Kibana will return warnings on startup for certain collectors if "duplicate documents" are found. Below are some examples:

```
server    log   [15:16:21.949] [warning][telemetry] Error scheduling task, received [task:oss_telemetry-vis_telemetry]: version conflict, document already exists (current version [3]): [version_conflict_engine_exception] [task:oss_telemetry-vis_telemetry]: version conflict, document already exists (current version [3]), with { index_uuid="4kgae7-8Rx6mWE5QAHhqkA" & shard="0" & index=".kibana_task_manager_1" }
server    log   [15:16:21.950] [debug][admin][elasticsearch][query] 409
```

```
server    log   [15:16:21.952] [warning][maps] Error scheduling telemetry task, received [task:Maps-maps_telemetry]: version conflict, document already exists (current version [3]): [version_conflict_engine_exception] [task:Maps-maps_telemetry]: version conflict, document already exists (current version [3]), with { index_uuid="4kgae7-8Rx6mWE5QAHhqkA" & shard="0" & index=".kibana_task_manager_1" }
```

### CODE REVIEW

**Comment by chrisronline** [comment]
This is interesting, and I'm eager to figure out the reason for all this flakiness. My worry with this particular step is it's not something a user would do when migrating from internal collection to Metricbeat collection, so if this is really a solution for what we're seeing, is it something that users will have to do as well, so telemetry data isn't lost?

**Comment by cachedout** [comment]
I don't think this bug has anything to do with whether one uses Metricbeat collection or internal collection but rather the order in which Kibana is started (and then subsequently restarted). 

My vote would be that we merge this because it doesn't appear to have anything to do with parity between the two collection methods which is what these tests are designed to enforce and then we also file an issue in the Kibana repo for further investigation regarding the warnings which appear in the log where we include the findings from this issue. WDYT?

**Comment by chrisronline** [comment]
> it doesn't appear to have anything to do with parity between the two collection methods which is what these tests are designed to enforce

I'm not sure I agree with this. 

We are seeing failures because the two documents are not identical. The reason why might be unrelated to this, but it's important to us that the documents are identical. If we merge this in, it feels like a bandaid that might make us think we solved this and don't need to reinvestigate.

I vote we open the ticket in Kibana and try and figure out exactly what's going on here before merging this.

**Comment by cachedout** [comment]
I have filed a bug on the Kibana side here: https://github.com/elastic/kibana/issues/51711

**Comment by chrisronline** [comment]
@cachedout What `ES_BUILD_URL` were you using when you consistently saw the failure in 7.4?

**Comment by cachedout** [comment]
I spoke to the telemetry folks this morning and learned a little more about what's going on here. The conversation [can be seen here](https://elastic.slack.com/archives/C3R6B8QV7/p1574844609295400).

The tl;dr is that this is a known issue which was described in https://github.com/elastic/kibana/issues/41827 but closed 13 days ago. The [PR which fixed it](https://github.com/elastic/kibana/pull/50232) went into master and was backported to 7.x but did not make its way into 7.4/7.5.

@gmmorris is investigating the feasibility of a backport into those branches presently.

**Comment by gmmorris** [comment]
The 7.5 backport #51796 has gone smoothly, and we should be able to merge it in if CI goes green.
The 7.4 backport #51797 ... less so. I think it's ok, I had to clean up some conflicts, so waiting to see how CI goes.

**Comment by cachedout** [comment]
Just to close the loop on next-steps, if elastic/kibana#51796 and elastic/kibana#51797 both get merged, we should wait a little while to ensure our tests start passing and then close this PR without merging it.

**Comment by cachedout** [comment]
> @cachedout What ES_BUILD_URL were you using when you consistently saw the failure in 7.4?

@chrisronline Apologies for not seeing this yesterday. The URL is `ES_BUILD_URL=snapshots.elastic.co/7.4.3-500ae70b`.

**Comment by cachedout** [comment]
OK, another update here. There is resistance to backporting the fix because there isn't another planned release of the 7.4 series and they don't want to backport to 7.5 so late in the release cycle.

I think that we should merge this PR since the problem is well-known on the Kibana side and they're managing it in release branches going forward.

**Comment by chrisronline** [comment]
Interesting! This is a great find.

So it sounds like the task manager is unable to start up a task (because it already exists) so when the telemetry collector goes to fetch data, it can't fetch anything because the task isn't even running? If so, this should be pretty easy to reproduce even without involving a separate collection mechanism, right? Like, this should affect internal collection too?

**Comment by cachedout** [comment]
> If so, this should be pretty easy to reproduce even without involving a separate collection mechanism, right? Like, this should affect internal collection too?

Yes. I believe this is correct. This will happen no matter the collection method and has everything to do with the task manager after a kibana restart.

**Comment by chrisronline** [comment]
> I think that we should merge this PR since the problem is well-known on the Kibana side and they're managing it in release branches going forward.

If they merge their PRs, we don't need this PR right?

**Comment by cachedout** [comment]
> If they merge their PRs, we don't need this PR right?

I do not believe they are planning on merging the above backport PRs. 

The PR for 7.4 is open but I believe it will be closed because [no further 7.4 releases are planned](https://docs.google.com/spreadsheets/d/1ohPWWkUy5Cc2k4HJSx3XBKXY5nxeuiehApFLFHdHM3U/edit#gid=899871603). 

The PR for 7.5 was closed because there was a desire expressed on the Kibana team not to merge it there because the risk was too high. That discussion can be seen here: https://elastic.slack.com/archives/C3R6B8QV7/p1574859209338900

**Comment by chrisronline** [comment]
> The PR for 7.4 is open but I believe it will be closed because no further 7.4 releases are planned.

Good point. This testing tool relies on release artifacts, which won't be generated without a planned release (afaik). I think they should merge back to `7.4` just in case (cc @gmmorris) there is a release, but we might have to apply the changes in this PR to `7.4` 

> The PR for 7.5 was closed because there was a desire expressed on the Kibana team not to merge it there because the risk was too high.

I can only imagine this is a temporary thing. They probably just want to freeze the branch until `7.5.0` GAs, then we should be able to freely merge into it. I did follow up in the slack conversation to verify

It sounds like we might only need to apply this fix to `7.4` though (which is good as I don't feel comfortable with the step in this PR as it's not something a user would ever do)

**Comment by cachedout** [comment]
>  This testing tool relies on release artifacts, which won't be generated without a planned release (afaik). I think they should merge back to 7.4 just in case (cc @gmmorris) there is a release, but we might have to apply the changes in this PR to 7.4

This brings up another point, which is that given that we're at the end of the line for 7.4 should we continue to maintain this testing branch or should we remove it in favor of the existing 7.5/7.x branch tests?

I think we should probably keep it around because, I assume, there could be an unplanned release in 7.4 like a security fix and if that happened, we'd want to keep test coverage around for so long as the release is officially supported, regardless of whether or not another release is planned but I would be very interested in what you think, @chrisronline .

**Comment by chrisronline** [comment]
There is no harm in leaving the tests, right? They'll only run when another release artifact is created so shouldn't be wasting resources or anything. 

I think we should merge this PR _only_ into 7.4 (we unmerge it if another release candidate in `7.4.x` is built _and_ https://github.com/elastic/kibana/pull/51797 is merged). WDYT?

**Comment by cachedout** [comment]
> I think we should merge this PR only into 7.4 (we unmerge it if another release candidate in 7.4.x is built and elastic/kibana#51797 is merged). WDYT?

My impression from the comments in Slack is that it still doesn't appear to me like they want this in 7.5. Specifically, Court mentioned: `I’m not sure what the impact is or what has changed to warrant a TM (task manager) change in a patch version`

 I am on board with limiting this to 7.4 though and then also to 7.5 if they don't want to do the backport there. Sound good?

**Comment by chrisronline** [comment]
Yes I agree. I think it will be backported to `7.5`, but if not, then we should merge this into our `7.5`

**Comment by gmmorris** [comment]
Following a conversation with Court we've decided to hold off on merging those backports as they are quite big enhancements to be released in a patch version.

**Comment by cachedout** [comment]
Thanks for letting us know, @gmmorris 

@chrisronline Given the above, I have opened a PR to cherry-pick this to 7.5

**Comment by chrisronline** [comment]
I'm not sure we should merge any of this (into any branch) at this point. I don't feel like we know enough about the _why_ here. I'd feel more comfortable merging this change if we truly understood why this step is necessary to fix the problem

**Comment by ycombinator** [comment]
I was away on break for a week so I've not been able to attend to this PR until now. Catching up on the conversation above, I'm gathering the following:

1. There was a fix needed in Kibana, that was merged into `master` via https://github.com/elastic/kibana/pull/50232 on 2019-11-14. With this fix in place, have we seen the `master` parity tests fail with the same failure as mentioned in this PR's description? Looking at the build history for https://internal-ci.elastic.co/job/elastic+estf-monitoring-snapshots+master+multijob-kibana/, it appears we've had a green `master` build for Kibana parity tests since 2019-11-14. To me that indicates that the fix made in https://github.com/elastic/kibana/pull/50232 is sufficient and this PR here (targeting `master` is no longer necessary).

2. Assuming the fix made in https://github.com/elastic/kibana/pull/50232 resolves the issue in `master`, backporting it to `7.x` should also fix the Kibana parity tests in `7.x`. It looks like the `7.x` backport PR (https://github.com/elastic/kibana/pull/50705) was merged on 2019-11-14. Looking at the build history for https://internal-ci.elastic.co/job/elastic+estf-monitoring-snapshots+7.x+multijob-kibana/, it appears we've had a green `7.x` build for Kibana parity tests since 2019-11-25, except on two occasions. On those two occasions the build jobs failed very early in their life and never got to even setting up the various stack products needed for parity testing. So, again, to me this indicates that the fix backported in https://github.com/elastic/kibana/pull/50705 is sufficient for `7.x` Kibana parity tests to be green.

3. Given that we're seeing success with the Kibana fixes in `master` (https://github.com/elastic/kibana/pull/50232) and `7.x` (https://github.com/elastic/kibana/pull/50705), I believe we should also backport the fix to `7.5` now. I understand we didn't want to do this close to the release deadline but now that the release it out, perhaps it's time to resurrect https://github.com/elastic/kibana/pull/51796 and merge it into `7.5`— @gmmorris?

4. Finally, for `7.4`, given that it's not an active branch any more, I _personally_ don't think we need to bother with any changes to it — either on the Kibana side or the parity tests side. At some point the parity tests branches will get updated (@liza-mae do you have plans to do this soonish for all ESTF tests now that 7.5.0 is released?) so the `7.4` build job for parity tests will go away anyway.

In short, it seems to me that the fix for this needed/needs to be in Kibana, not in parity testing code. So I'm +1 for closing this PR and its backports unmerged and resurrecting https://github.com/elastic/kibana/pull/51796 and merging it into `7.5`. Thoughts?

**Comment by cachedout** [comment]
> To me that indicates that the fix made in elastic/kibana#50232 is sufficient and this PR here (targeting master) is no longer necessary.

I agree that no matter the fate of the Kibana backports, this particular PR should not remain open so I am closing it.

**Comment by cachedout** [comment]
Just to circle back around on this, it appears that there is still resistance to putting https://github.com/elastic/kibana/pull/51796 into 7.5 and we are now waiting on the telemetry team to possibly provide a workaround.

**Comment by chrisronline** [comment]
Thanks for bump here @cachedout.

I guess I'm still struggling to understand _how_ the PR actually helps. From conversations in slack, it sounds like the warning we are seeing (from the PR description) are actually innocuous  (cc @gmmorris to confirm) so I don't understand how the changes in his PR will affect our parity tests. Do you mind giving me a rundown of how it's connected so I can fully understand?

Thanks!

**Comment by cachedout** [comment]
@chrisronline Sure thing!

### Summary

What's happening here is that when Kibana starts up for the first time and there is no Task Manager index in existence, for a very brief time during the startup phase, the `usage` dictionary in the API response will contain certain keys, such as `visualization_types`. After the tasks manager has been fully initialized, those keys are no longer present.

Here is how to reproduce this on a 7.4 stack. I recommend you read all the steps and prepare your environment before testing since it requires a few operations in rapid succession:

1. Ensure that you are starting with a fresh version of Elasticsearch. If you are not, or you wish to reset your test environment, ensure that there are no indices present matching the `.kibana*` pattern.

2. After starting Elasticsearch and ensuring the environment is as-described in step 1, start Kibana. Move very quickly to a window where you can perform step 3.

3. Issue an HTTP request to Kibana to the `api/stats?extended=true`. It may take a few moments as Kibana starts to initialize before this endpoint is responsive. You want to issue a request to this endpoint as quickly as possible before the Task Manager is fully initialized.

4. Note that if you catch this fast enough, you will see a key in the `usage` field for `visualization_types` with a value of an empty dictionary.

5. Re-issue the request a second time after waiting just a few seconds. Note that the reply no longer contains the `visualization_types` key in the `usage` dictionary.

TBH, I'm not positive about what is _supposed_ to be in the `visualization_types` field. It looks like it is querying the task manager for the presence of a collection task itself but I'm not sure if there is supposed to be something else there. But based on what I've discovered so far, that appears to me to be the case but the intended behavior is murky.

## Effects on parity testing

While this may be an edge-case bug in telemetry, the effects seem limited to the very first moments of a Kibana's lifecycle. I can find no indication that there is a difference between the way Metricbeat handles this situation versus the way that internal collection does. If we were to switch the order of collection steps in the tests, the failures should switch from one collection method to another as well. I have not modified the test suite to test this but am happy to if you think it would be good to verify.

## Next steps

I think it would be good to learn a little more from @tsullivan about the intended interaction between the task manager and these telemetry tasks and to understand whether or not there is a true effect on what telemetry data is ultimately collected and shipped.

**Comment by chrisronline** [comment]
Okay cool, I'm also seeing this.

My best guess here is [this isReady check](https://github.com/elastic/kibana/blob/7.4/x-pack/legacy/plugins/oss_telemetry/server/lib/collectors/visualizations/get_usage_collector.ts#L57) isn't completely accurate - it should be waiting for whatever the consistent response is.

The first couple responses from the task manager request are:
```
[{"taskType":"vis_telemetry","state":{"stats":{},"runs":0},"params":{},"attempts":1,"scheduledAt":"2019-12-06T15:25:05.352Z","startedAt":"2019-12-06T15:25:08.740Z","retryAt":"2019-12-06T15:35:08.740Z","runAt":"2019-12-06T15:25:05.352Z","status":"running","id":"oss_telemetry-vis_telemetry"}]
```

Then, it changes to:
```
[{"taskType":"vis_telemetry","state":{"runs":1},"params":{},"attempts":1,"scheduledAt":"2019-12-06T15:25:05.352Z","startedAt":null,"retryAt":null,"runAt":"2019-12-07T05:00:00.000Z","status":"idle","id":"oss_telemetry-vis_telemetry"}]
```

I think the key here is the `runs`. It looks like [our isReady code](https://github.com/elastic/kibana/blob/master/x-pack/legacy/plugins/oss_telemetry/server/lib/collectors/visualizations/get_usage_collector.ts#L13) should be checking against `state.runs > 0` before considering the telemetry data ready.

@tsullivan WDYT?

**Comment by chrisronline** [comment]
I opened https://github.com/elastic/kibana/pull/52446 where we discuss this fix
